File: libs/language/src/lib/utils/NodeExtender.ts
import { isNonNullable } from '@blue-labs/shared-utils';
import { NodeProvider } from '../NodeProvider';
import { BlueNode } from '../model';
import { NodeProviderWrapper } from './NodeProviderWrapper';
import { CORE_TYPE_BLUE_IDS } from './Properties';
import { Limits } from './limits/Limits';
import { Nodes } from './Nodes';

/**
 * Strategies for handling missing elements
 */
export type MissingElementStrategy = 'THROW_EXCEPTION' | 'RETURN_EMPTY';

/**
 * Class for extending nodes with their resolved references
 */
export class NodeExtender {
  private nodeProvider: NodeProvider;
  private strategy: MissingElementStrategy;

  /**
   * Creates a new NodeExtender with the specified NodeProvider and optional strategy
   * @param nodeProvider - The NodeProvider to use for resolving nodes
   * @param strategy - The strategy to use for missing elements (defaults to THROW_EXCEPTION)
   */
  constructor(nodeProvider: NodeProvider, strategy?: MissingElementStrategy) {
    this.nodeProvider = NodeProviderWrapper.wrap(nodeProvider);
    this.strategy = strategy || 'THROW_EXCEPTION';
  }

  /**
   * Extends a node with its resolved references
   * @param node - The node to extend
   * @param limits - The limits to apply when extending
   */
  public extend(node: BlueNode, limits: Limits): void {
    this.extendNode(node, limits, '');
  }

  private extendNode(
    currentNode: BlueNode,
    currentLimits: Limits,
    currentSegment: string,
    skipLimitCheck = false,
  ): void {
    if (!skipLimitCheck) {
      if (!currentLimits.shouldExtendPathSegment(currentSegment, currentNode)) {
        return;
      }

      currentLimits.enterPathSegment(currentSegment, currentNode);
    }

    try {
      const blueId = currentNode.getBlueId();
      if (
        blueId &&
        Nodes.hasBlueIdOnly(currentNode) &&
        !CORE_TYPE_BLUE_IDS.includes(
          blueId as (typeof CORE_TYPE_BLUE_IDS)[number],
        )
      ) {
        const resolvedNodes = this.fetchNode(currentNode);
        if (resolvedNodes && resolvedNodes.length > 0) {
          const preserveReferenceBlueId = this.shouldPreserveReferenceBlueId(
            currentSegment,
            skipLimitCheck,
          );
          if (resolvedNodes.length === 1) {
            const resolvedNode = resolvedNodes[0];
            this.mergeNodes(currentNode, resolvedNode, preserveReferenceBlueId);
          } else {
            const mergedNodes = resolvedNodes.map((node) => node.clone());
            const listNode = new BlueNode().setItems(mergedNodes);
            this.mergeNodes(currentNode, listNode, preserveReferenceBlueId);
          }
        }
      }

      const typeNode = currentNode.getType();
      if (typeNode) {
        this.extendNode(typeNode, currentLimits, 'type', true);
      }
      const itemTypeNode = currentNode.getItemType();
      if (itemTypeNode) {
        this.extendNode(itemTypeNode, currentLimits, 'itemType', true);
      }
      const keyTypeNode = currentNode.getKeyType();
      if (keyTypeNode) {
        this.extendNode(keyTypeNode, currentLimits, 'keyType', true);
      }
      const valueTypeNode = currentNode.getValueType();
      if (valueTypeNode) {
        this.extendNode(valueTypeNode, currentLimits, 'valueType', true);
      }

      const properties = currentNode.getProperties();
      if (properties) {
        Object.entries(properties).forEach(([key, value]) => {
          this.extendNode(value, currentLimits, key, false);
        });
      }

      const items = currentNode.getItems();
      if (items && items.length > 0) {
        for (let i = 0; i < items.length; i++) {
          this.extendNode(items[i], currentLimits, String(i), false);
        }
      }
    } finally {
      if (!skipLimitCheck) {
        currentLimits.exitPathSegment();
      }
    }
  }

  private fetchNode(node: BlueNode): BlueNode[] | null {
    const nodeBlueId = node.getBlueId();
    if (!nodeBlueId) {
      if (this.strategy === 'RETURN_EMPTY') {
        return null;
      } else {
        throw new Error(`No blueId found for node: ${node.getName()}`);
      }
    }

    const resolvedNodes = this.nodeProvider.fetchByBlueId(nodeBlueId);
    if (!resolvedNodes || resolvedNodes.length === 0) {
      if (this.strategy === 'RETURN_EMPTY') {
        return null;
      } else {
        throw new Error(`No content found for blueId: ${node.getBlueId()}`);
      }
    }
    return resolvedNodes;
  }

  private shouldPreserveReferenceBlueId(
    currentSegment: string,
    skipLimitCheck: boolean,
  ): boolean {
    return (
      skipLimitCheck &&
      ['type', 'itemType', 'keyType', 'valueType'].includes(currentSegment)
    );
  }

  private mergeNodes(
    target: BlueNode,
    source: BlueNode,
    preserveReferenceBlueId: boolean,
  ): void {
    if (!preserveReferenceBlueId) {
      target.setReferenceBlueId(undefined);
    }
    target.setName(source.getName());
    target.setDescription(source.getDescription());
    target.setType(source.getType()?.clone());
    target.setItemType(source.getItemType()?.clone());
    target.setKeyType(source.getKeyType()?.clone());
    target.setValueType(source.getValueType()?.clone());
    const sourceValue = source.getValue();
    if (isNonNullable(sourceValue)) {
      target.setValue(sourceValue);
    }
    target.setItems(source.getItems()?.map((item) => item.clone()));
    const sourceProperties = source.getProperties();
    target.setProperties(
      sourceProperties === undefined
        ? undefined
        : Object.fromEntries(
            Object.entries(sourceProperties).map(([key, value]) => [
              key,
              value.clone(),
            ]),
          ),
    );
    target.setBlue(source.getBlue()?.clone());
  }
}

File: references/java-sdk/src/main/java/blue/language/utils/NodeExtender.java
package blue.language.utils;

import blue.language.NodeProvider;
import blue.language.model.Node;
import blue.language.utils.limits.Limits;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static blue.language.utils.Properties.CORE_TYPE_BLUE_IDS;

public class NodeExtender {

    public enum MissingElementStrategy {
        THROW_EXCEPTION,
        RETURN_EMPTY
    }

    private final NodeProvider nodeProvider;
    private final MissingElementStrategy strategy;

    public NodeExtender(NodeProvider nodeProvider) {
        this(nodeProvider, MissingElementStrategy.THROW_EXCEPTION);
    }

    public NodeExtender(NodeProvider nodeProvider, MissingElementStrategy strategy) {
        this.nodeProvider = NodeProviderWrapper.wrap(nodeProvider);
        this.strategy = strategy;
    }

    public void extend(Node node, Limits limits) {
        extendNode(node, limits, "");
    }

    private void extendNode(Node currentNode, Limits currentLimits, String currentSegment) {
        extendNode(currentNode, currentLimits, currentSegment, false);
    }

    private void extendNode(Node currentNode, Limits currentLimits, String currentSegment, boolean skipLimitCheck) {
        if (!skipLimitCheck) {
            if (!currentLimits.shouldExtendPathSegment(currentSegment, currentNode)) {
                return;
            }

            currentLimits.enterPathSegment(currentSegment, currentNode);
        }

        try {
            if (currentNode.getBlueId() != null && !CORE_TYPE_BLUE_IDS.contains(currentNode.getBlueId())) {
                List<Node> resolvedNodes = fetchNode(currentNode);
                if (resolvedNodes != null && !resolvedNodes.isEmpty()) {
                    if (resolvedNodes.size() == 1) {
                        Node resolvedNode = resolvedNodes.get(0);
                        mergeNodes(currentNode, resolvedNode);
                    } else {
                        List<Node> mergedNodes = resolvedNodes.stream()
                                .map(Node::clone)
                                .collect(Collectors.toList());
                        Node listNode = new Node().items(mergedNodes);
                        mergeNodes(currentNode, listNode);
                    }
                }
            }

            // Handle type nodes
            if (currentNode.getType() != null) {
                extendNode(currentNode.getType(), currentLimits, "type", true);
            }
            if (currentNode.getItemType() != null) {
                extendNode(currentNode.getItemType(), currentLimits, "itemType", true);
            }
            if (currentNode.getKeyType() != null) {
                extendNode(currentNode.getKeyType(), currentLimits, "keyType", true);
            }
            if (currentNode.getValueType() != null) {
                extendNode(currentNode.getValueType(), currentLimits, "valueType", true);
            }

            Map<String, Node> properties = currentNode.getProperties();
            if (properties != null) {
                properties.forEach((key, value) -> {
                    extendNode(value, currentLimits, key, false);
                });
            }

            List<Node> items = currentNode.getItems();
            if (items != null && !items.isEmpty()) {
                reconstructList(items);
                for (int i = 0; i < items.size(); i++) {
                    extendNode(items.get(i), currentLimits, String.valueOf(i), false);
                }
            }
        } finally {
            if (!skipLimitCheck) {
                currentLimits.exitPathSegment();
            }
        }
    }

    private void reconstructList(List<Node> items) {
        while (!items.isEmpty()) {
            Node firstItem = items.get(0);
            String blueId = firstItem.getBlueId();
            if (blueId == null) {
                break;
            }
            List<Node> resolved = nodeProvider.fetchByBlueId(blueId);
            if (resolved == null || resolved.size() == 1) {
                break;
            }
            items.remove(0);
            items.addAll(0, resolved);
        }
    }

    private List<Node> fetchNode(Node node) {
        List<Node> resolvedNodes = nodeProvider.fetchByBlueId(node.getBlueId());
        if (resolvedNodes == null || resolvedNodes.isEmpty()) {
            if (strategy == MissingElementStrategy.RETURN_EMPTY) {
                return null;
            } else {
                throw new IllegalArgumentException("No content found for blueId: " + node.getBlueId());
            }
        }
        return resolvedNodes;
    }

    private void mergeNodes(Node target, Node source) {
        target.name(source.getName());
        target.description(source.getDescription());
        target.type(source.getType());
        target.itemType(source.getItemType());
        target.keyType(source.getKeyType());
        target.valueType(source.getValueType());
        target.value(source.getValue());
        target.items(source.getItems());
        target.properties(source.getProperties());
        target.constraints(source.getConstraints());
    }
}